src/components/Web3Provider.js

```diff
@@ -40,29 +55,45 @@ const Web3Provider = ({ children }) => {
   useEffect(() => {
     if (sdk) {
       sdk.addListeners({
-        onSafeInfo: setSafeInfo,
+        onSafeInfo: handleSafeInfo,
       });
 
+      if (safeInfo) {
+        setStatus("SUCCESS");
+      }
+
       return () => {
         sdk.removeListeners();
       };
     }
-  }, [sdk]);
+  }, [sdk, safeInfo, handleSafeInfo]);
 
   const handleAsyncInit = useCallback(() => {
     handleInit();
   }, [handleInit]);
 
-  const handleGetArtifact = useCallback(async (contractName) => {
-    let contractArtifact = contractArtifacts[contractName];
-    if (!contractArtifact) {
+  const handleGetArtifact = useCallback(async (dirtyContractName) => {
+    const contractName = dirtyContractName.replace(/\..*/, "");
+    if (!globalArtifacts[contractName]) {
       // Load from default folder with contractName.json
-      const response = await fetch(`contracts/${contractName}.json`);
-      contractArtifact = await response.json();
-      contractArtifacts[contractName] = contractArtifact;
+      globalArtifactPromises[contractName] = readContractJSON(contractName);
+
+      // Await the result to make sure it gets added the globalArtifacts
+      globalArtifacts[contractName] = await globalArtifactPromises[
+        contractName
+      ];
```

Comment:
Otherwise when we await the same contract multiple times, multiple requests will be made. If we cache the promise instead, it will just await the one request multiple times.